text/0000-stable-simd.md

```diff
@@ -0,0 +1,720 @@
+
+Note that AVX-512 names are missing from this list, but that's because we
+haven't implemented any AVX-512 intrinsics yet. Those'll get stabilized on their
+own once implemented. Additionally note that `mmx` is missing from this list.
+For reasons discussed later, it's proposed that MMX types are omitted from the
+first pass of stabilization. AMD also has some specific features supported
+(`sse4a`, `tbm`), and so do ARM, MIPS, and PowerPC, but none of these feature
+names a proposed for becoming stable in the first pass.
+
+## The `target_feature` value in `#[cfg]`
+
+In addition to enabling target features for a function the compiler will also
+allow statically testing whether a particular target feature is enabled. This
+corresponds to the `cfg_target_feature` feature today in rustc, and can be seen
+via:
+
+```rust
+#[cfg(target_feature = "avx")]
+fn foo() {
+    // implementation that can use `avx`
+}
+
+#[cfg(not(target_feature = "avx"))]
+fn foo() {
+    // a fallback implementation
+}
+```
+
+Additionally this is also made available to `cfg!`:
+
+```rust
+if cfg!(target_feature = "avx") {
+    println!("this program was compiled with AVX support");
+}
+```
+
+The `#[cfg]` attribute and `cfg!` macro statically resolve and **do not do
+runtime dispatch**. Tweaking these functions is currently done via the `-C
+target-feature` flag to the compiler. This flag to the compiler accepts a
+similar set of strings to the ones specified above and is already "stable".
+
+## The `cfg_feature_enabled!` Macro
+
+One mode of operation with vendor intrinsics is to compile *part* of a program
+with certain CPU features enabled but not the entire program. This way a
+portable program can be compiled which runs across a broad range of hardware
+which can still benefit from optimized implementations for particular hardware
+at runtime.
+
+The crux of this support in libstd is this macro provided by libstd,
+`cfg_feature_enabled!`. The macro will accept one argument, a string literal.
+The string can be any feature passed to `#[target_feature(enable = ...)]` for
+the platform you're compiling for. Finally, the macro will resolve to a `bool`
+result.
+
+For example on x86 you could write:
+
+```rust
+if cfg_feature_enabled!("sse4.1") {
+    println!("this cpu has sse4.1 features enabled!");
+}
+```
+
+It would, however, be an error to write this on x86 cpus:
+
+```rust
+cfg_feature_enabled!("neon"); //~ ERROR: neon is an ARM feature, not x86
+cfg_feature_enabled!("foo"); //~ ERROR: unknown target feature for x86
+```
+
+The macro is intended to be implemented in the `std` crate (**not** `core`) and
+made available via the normal macro preludes. The implementation of this macro
+is expected to be what [`stdsimd`][stdsimd] does today, notably:
+
+* The first time the macro is invoked all the local CPU features will be
+  detected.
+* The detected features will then be cached globally (currently in a bitset) for
+  the rest of the execution of the program.
```

Comment:
In general, availability of instructions can change in runtime, so this cache can be invalidated, but this should be relevant only for bare-metal environments.

For example, availability of AVX (https://github.com/rust-lang-nursery/stdsimd/pull/154) in addition to `CPUID` depends on bits in `XCR0` that can be changed by privileged software.
(Fun fact: even some `CPUID` bits can change in runtime, e.g. `MONITOR` bit reporting availability of `_mm_monitor`/`_mm_mwait`).

Reply:
@petrochenkov The RFC only provides run-time support in `std`. Users of `core` should choose whatever run-time detection system fits their code. If they are writing user space code, they can choose a library that provides behavior similar to the run-time facilities in `std`. If they are writing kernel code, they will need something else, but they can built it on top of crates that already do `cpuid` queries like `cupid`.